cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,318 @@
+  private static final Logger LOG = LoggerFactory.getLogger(WorkflowDataset.class);
+  private static final Type ACTION_RUNS_TYPE = new TypeToken<List<ActionRuns>>() { }.getType();
+
+  private final Table table;
+
+  WorkflowDataset(Table table) {
+    super("ignored", table);
+    this.table = table;
+  }
+
+  private byte[] createRowKey(String key, long time) {
+    byte[] result = Bytes.toBytes(key);
+    // Size of long is being added to the row key since the start time is being added at the end
+    byte[] rowKey = new byte[result.length + Bytes.SIZEOF_LONG];
+    int offset = Bytes.putBytes(rowKey, 0, result, 0, result.length);
+    Bytes.putBytes(rowKey, offset, Bytes.toBytes(time), 0, Bytes.SIZEOF_LONG);
+    return rowKey;
+  }
+
+  public void write(Id.Program id, RunRecordMeta runRecordMeta, List<ActionRuns> actionRunsList) {
+
+    long start = runRecordMeta.getStartTs();
+    String key = id.getApplication().getNamespaceId() + DELIMITER
+      + id.getApplicationId() + DELIMITER + id.getId() + DELIMITER;
+    byte[] rowKey = createRowKey(key, start);
+
+    long timeTaken = runRecordMeta.getStopTs() - start;
+
+    String value = GSON.toJson(actionRunsList, ACTION_RUNS_TYPE);
+
+    table.put(rowKey, RUNID, Bytes.toBytes(runRecordMeta.getPid()));
+    table.put(rowKey, TIME_TAKEN, Bytes.toBytes(timeTaken));
+    table.put(rowKey, NODES, Bytes.toBytes(value));
+  }
+
+  public List<WorkflowRunRecord> scan(Id.Program id, long timeRangeStart, long timeRangeEnd) {
+    String namespaceId = id.getApplication().getNamespaceId();
+    String applicationId = id.getApplicationId();
+    String workflowId = id.getId();
+
+    String key = namespaceId + DELIMITER + applicationId + DELIMITER + workflowId + DELIMITER;
+    byte[] startRowKey = createRowKey(key, timeRangeStart);
+    byte[] endRowKey = createRowKey(key, timeRangeEnd);
+    Scan scan = new Scan(startRowKey, endRowKey);
+
+    Scanner scanner = table.scan(scan);
+    Row indexRow;
+    List<WorkflowRunRecord> workflowRunRecordList = new ArrayList<>();
+    while ((indexRow = scanner.next()) != null) {
+      Map<byte[], byte[]> columns = indexRow.getColumns();
+      String workflowRunId = Bytes.toString(columns.get(RUNID));
+      long timeTaken = Bytes.toLong(columns.get(TIME_TAKEN));
+
+      List<ActionRuns> actionRunsList = GSON.fromJson(Bytes.toString(columns.get(NODES)), ACTION_RUNS_TYPE);
+      WorkflowRunRecord workflowRunRecord = new WorkflowRunRecord(workflowRunId, timeTaken, actionRunsList);
+      workflowRunRecordList.add(workflowRunRecord);
+    }
+    return workflowRunRecordList;
+  }
+
+  public WorkflowDataset.BasicStatistics statistics(Id.Workflow id, long startTime,
+                                                    long endTime, List<Double> percentiles) {
+
+    List<WorkflowRunRecord> workflowRunRecords = scan(id, startTime, endTime);
+    int count = workflowRunRecords.size();
+
+    if (count == 0) {
+      return null;
```

Comment:
If this method returns null, it should be marked ``@Nullable``. However, should we return an empty statistics object instead?